cpp/include/cudf/merge.hpp

```diff
@@ -0,0 +1,281 @@
+ */
+
+#pragma once
+
+#include <vector>
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+
+#include <cudf/table/row_operators.cuh>
+
+
+namespace { // anonym.
+/**
+ * @brief Source table identifier to copy data from.
+ */
+enum class side : bool { LEFT, RIGHT };
+
+  struct X{};
+
+  template<typename T>
+  struct compile_time_printer
+  {
+    compile_time_printer(void) = delete;
+  };
+
+  //only the T == X full specialization
+  //can be instantiated:
+  //
+  template<>
+  struct compile_time_printer<X>
+  {
+    compile_time_printer(void){}
+  };
+  
+} //anonym.
+
+namespace cudf {
+namespace experimental {
+/**
+ * @brief tagged_element_relational_comparator uses element_relational_comparator to provide "tagged-index" comparation logic.
+ *
+ * Special treatment is necessary in several thrust algorithms (e.g., merge()) where 
+ * the index affinity to the side is not guaranteed; i.e., the algorithms rely on 
+ * binary functors (predicates) where the operands may transparently switch sides.
+ *
+ * For example, 
+ *         thrust::merge(left_container,
+ *                       right_container,
+ *                       predicate(lhs, rhs){...});
+ *         can create 4 different use-cases, inside predicate(...):
+ *
+ *         1. lhs refers to the left container; rhs to the right container;
+ *         2. vice-versa;
+ *         3. both lhs and rhs actually refer to the left container;
+ *         4. both lhs and rhs actually refer to the right container;
+ *
+ * Because of that, one cannot rely on the predicate having *fixed* references to the containers.
+ * Each invocation may land in a different situation (among the 4 above) than any other invocation.
+ * Also, one cannot just manipulate lhs, rhs (indices) alone; because, if predicate always applies
+ * one index to one container and the other index to the other container, 
+ * switching the indices alone won't suffice in the cases (3) or (4), 
+ * where the also the containers must be changed (to just one instead of two) 
+ * independently of indices;
+ *
+ * As a result, a special comparison logic is necessary whereby the index is "tagged" with side information
+ * and consequently comparator functors (predicates) must operate 
+ * on these tagged indices rather than on raw indices.
+ *  
+ */
+template <typename ExtractorT,
+          bool has_nulls = true>
+struct tagged_element_relational_comparator {
+
+  __host__ __device__
+  tagged_element_relational_comparator(column_device_view lhs,
+                                       column_device_view rhs,
+                                       null_order null_precedence)
+    : lhs{lhs}, rhs{rhs}, null_precedence{null_precedence}
+  {
+  }
+
+
+  template <typename Element,
+            typename TaggedIndexT,
+            std::enable_if_t<cudf::is_relationally_comparable<
+                               Element, Element>()>* = nullptr>
+  __device__
+  weak_ordering operator()(TaggedIndexT lhs_tagged_index,
```

Comment:
Why not just do @trevorsm7's idea and use strongly typed indices? You'd then need to provide four `operator()` overloads:
```
operator()(lhs_index a, rhs_index b)
operator()(lhs_index a, lhs_index b)
operator()(rhs_index a, lhs_index b)
operator()(rhs_index a, rhs_index b)
```

Providing the "side" information as runtime instead of compile time info will be slower and complicates the implementation.

Reply:
@jrhemstad, we've been through this numerous times. Because it doesn't work.

Specifically, `merge()` requires copy constructor between the strongly typed indices. It doesn't compile without it. And that cancels out the purpose of using strongly typed indices. Again, I've tried it. It doesn't work.

Detailed conversations here:
          1. https://nvidia.slack.com/archives/CDTANRCTT/p1572987546442300
          2. https://nvidia.slack.com/archives/C7MFRGM6H/p1572987322022500
          3. https://nvidia.slack.com/archives/C7MFRGM6H/p1572994428027500
          4. https://nvidia.slack.com/archives/CDTANRCTT/p1574364256110700 (probably most important, so far)

Reply:
https://nvidia.slack.com/archives/CDTANRCTT/p1574371129116400?thread_ts=1574364256.110700&cid=CDTANRCTT

Reply:
Yeah, this makes sense for merge since both input iterators need to be convertible to the output iterator, and once in the output iterator you have mixed lhs and rhs indices that you need to distinguish between.

Reply:
Maybe it would be possible to combine both approaches by having strongly typed iterators that can't be converted to each other, but _can_ be implicitly converted to the tagged index type? That might give you the compile-time dispatch on the comparator while still giving you the tagged output.

The tagged index type makes sense in the ColumnMerger implementation too where you branch inside the transform. This looks like a specialization of gather for two inputs.

Hm, I wonder if you could have avoided all this trouble by vertically concatenating the columns such that you'd only have had to deal with two offsets into the same column for the merge and then just using gather to build the assemble the output.

Reply:
To recap. (so that we start from same baseline understanding of what's going on):
thrust::merge(left_container,
                        right_container,
                        predicate(lhs, rhs){...});
          can create 4 different use-cases, inside predicate(...):

          1. lhs refers to the left container; rhs to the right container;
          2. vice-versa;
          3. both lhs and rhs actually refer to the left container;
          4. both lhs and rhs actually refer to the right container;

Because of that, one cannot rely on the predicate having *fixed* references to the containers; each invocation may land in a different situation (among the 4 above) than any other invocation;

Also, one cannot just manipulate lhs, rhs (indices) alone; because, if predicate always applies one of them to one container and the other to the other container, switching the indices won't help in the cases (3) or (4);

As a result, the tagged logic propagates through `predicate` inside `row_lexicographic_comparator` and then into `element_relational_comparator` which does the "sleight of hand" of pointing to the appropriate containers (columns). I don't see how the "strongly typed different counting iterators" can help. Especially, since I've tried it before and did _not_ work with `merge`.

One other way this could work (tried it in a simpler repro and it does work) is to use disjoint set of row indices for `left` and `right`; i.e., `left` uses row indices from in `(0..n-1)` and `right` uses row indices `(n..n+m-1)` (where `n` = number of `left` rows; `m` = number of `right rows). That has the effect of including the "tag" information of the indices in their values, directly, without adding an additional "side" type. 

I'm not sure if that brings much benefit (it certainly is less readable), because you'd still need to use the "sleight of hand" in a specialized version of `element_relational_comparator`. IOW, I don't see a way around using specialized versions of `element_relational_comparator` and `row_lexicographic_comparator` (or parameterizing `row_lexicographic_comparator` by an additional `relational_comparator` type).

Reply:
If you merged the columns vertically (possibly with an iterator, not actually copying into the same contiguous buffer... I guess you'd need an iterator instead of a column_device_view), then you could use (n..n+m-1) for the rhs indices without modifying the comparator at all

Reply:
Before actual merging of the columns I need the correct order of the indices (either tagged or disjoint...complementary is a better word). So, I need to `merge` correct indices based on a lexicographic sort of several columns. 

How would an iterator help me bypass using "special" versions of `row_lexicographic_comparator` and (consequently) `element_relational_comparator`? I don't understand what are you suggesting...

Reply:
`std::merge/thrust::merge` have an odd requirement that the two input ranges need to both be convertible to the lhs/rhs arguments to the binary predicate.

Example: https://wandbox.org/permlink/H4y6TCzmcgkdY7FK

`std::merge` makes this explicit, but thrust::merge was lacking the same documentation:
 > The types Type1 and Type2 must be such that objects of types InputIt1 and InputIt2 can be dereferenced and then implicitly converted to both Type1 and Type2.

As such, I don't see a way to avoid the additional runtime checks in the binary predicate. 

@trevorsm7's idea to logically concatenate the columns is interesting, but I don't see how to make that work without materializing the concatenated columns or putting runtime checks in the iterator, e.g., 
```
   if( i < n) // use top half
   else // use bottom half
```